### PR TITLE
Fix set geometrypath issue 2418

### PR DESCRIPTION
**Pull request by aymanhab**
Fixes issue #2418

### Brief summary of changes
Do not prepend Componentpath to socket connectees if path is already valid

### Testing I've completed
Ran all test cases and tested script in issue and produced correct socket connectees

### Looking for feedback on...
Not sure which test case to beef up to handle this change since it's generic and it can't be reproduced using a single component.

### CHANGELOG.md (choose one)

- no need to update because...
- updated...

The Doxygen for this PR can be viewed at http://myosin.sourceforge.net/?C=N;O=D; click the folder whose name is this PR's number.

---
This change is [Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/2504)

### CODE REVIEW

**Comment by aymanhab** [comment]
As of now this is PR is a place holder for discussion regarding the current behavior, no fix to merge yet.

**Comment by aymanhab** [comment]
@chrisdembia The ComponentPath  for the PathSpring sockets gets corrupted during call to prependComponentPathToConnecteePath(*force); which you introduced as part of  commit 7b182f5eab5341af24b98e776d33912a5718bae7 (delay_connectee_name). The code in question is triggered only if "connecteepath.isAbsolute()" is true.  Is there any documentation as to how this flag is supposed to be maintained?

**Comment by chrisdembia** [comment]
There is definitely a bug in the prepend code. For example, https://github.com/opensim-org/opensim-core/issues/2388

> Is there any documentation as to how this flag is supposed to be maintained?

I'm not sure I understand the question.

**Comment by aymanhab** [comment]
Thanks @chrisdembia Bug #2388 is a duplicate of the same bug I'm looking at #2418, and I'm trying to understand what are the assumptions that the method prependComponentPathToConnecteePath is making about the added component. If you don't recall that's fine, I will keep digging. Thank you

**Comment by aymanhab** [comment]
I believe the logic in the function prependComponentPathToConnecteePath should apply iff connections are within the added component/subcomponents (in addition to Absolute being true) and not otherwise.  I don't know how to check that connections are within the component so I check that the connection is not available from the Root instead. This fixes the bug and all test cases pass except for testComponentInterface which I'm investigating

**Comment by aymanhab** [comment]
@chrisdembia and @aseth1 when you have a chance please let me know if you see something wrong with the change in Component.cpp. I'm debugging testComponentInterface to see why it fails and if the assumptions in it were justified.

File: OpenSim/Common/Component.cpp [hunk 1]
```diff
@@ -163,12 +163,15 @@ void Component::addComponent(Component* subcomponent)
...
     for (auto& comp : subcomponent.updComponentList()) {
         for (auto& it : comp._socketsTable) {
-            it.second->prependComponentPathToConnecteePath(compPath);
+            if (!root.hasComponent(it.second->getConnecteePath()))
+                it.second->prependComponentPathToConnecteePath(compPath);
         }
         for (auto& it : comp._inputsTable) {
-            it.second->prependComponentPathToConnecteePath(compPath);
+            if (!root.hasComponent(it.second->getConnecteePath()))
+                it.second->prependComponentPathToConnecteePath(compPath);
```
Comment (chrisdembia) [inline OpenSim/Common/Component.cpp]: The connectees of inputs are not components.

Comment (aymanhab) [reply OpenSim/Common/Component.cpp]: Thanks @chrisdembia I undid the change for inputs, now all test cases pass.  Thank you :+1

**Comment by aymanhab** [comment]
@jimmyDunne When you have a chance you can test the artifact to verify that the issue has been fixed.

**Comment by aymanhab** [comment]
@chrisdembia this ended up being a two liner defensive fix, so should be easy to review when you have a chance.
@jimmyDunne Please let me know if you can test the fix using the build artifact. Thank you.

**Comment by jimmyDunne** [comment]
I built locally and tested using the code in #2418. I didn't get the exception and the PropertyByName() returns the correct path. So LGTM!

**Comment by aymanhab** [comment]
Thanks @jimmyDunne and @chrisdembia 👍

**Pull request merged**